envoy/config/xds_resources_delegate.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/optref.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/subscription.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * An interface for hooking into xDS resource fetch and update events.
+ * Currently, this interface only supports the SotW (state-of-the-world) xDS protocol.
+ */
+class XdsResourcesDelegate {
+public:
+  virtual ~XdsResourcesDelegate() = default;
+
+  /**
+   * Returns a list of xDS resources for the given authority and type. It is up to the
+   * implementation to determine what resources to supply, if any.
+   *
+   * This function is intended to only be called on xDS fetch startup, and allows the
+   * implementation to return a set of resources to be loaded and used by the Envoy instance
+   * (e.g. persisted resources in local storage).
+   *
+   * @param authority_id Unique id for the authority/control plane from which the resources are
+   *   obtained. Typically either a cluster name or a host name.
+   * @param resource_type_url The URL for the type of the resource (e.g. Secrets, Clusters, etc).
+   * @return A set of xDS resources for the given authority and type.
+   */
+  virtual std::vector<envoy::service::discovery::v3::Resource>
+  getResources(const std::string& authority_id, const std::string& resource_type_url) PURE;
+
+  /**
+   * Invoked when SotW xDS configuration updates have been received from an xDS authority, have been
```

Comment:
I'm trying to think whether what we want here is a mechanism that's called on each update, or just a way to serialize the state (config).
On the one-hand I can see the simplicity in serializing the current state and loading it when the Envoy restarts (albeit some serialization work will be required). On the other-hand saving the config makes this more resilient to binary upgrades and makes saving deltas possible.
WDYT?

Reply:
Adi & I chatted a bit about this.  A problem with the way Envoy handles xDS now is that there is no unified notion of the state of the dynamic config that can be serialized and re-loaded.  We receive xDS resources from a control plane and play those updates to the various consumers (e.g. SdsApi, CdsApi, etc).  This makes it harder to handle the Delta protocol with this mechanism, though it works reasonably well for SotW.  Since our initial intention with this extension point was to support Envoy Mobile use cases, we feel okay with limiting the feature to SotW for now.

Would like to know more senior maintainers thoughts on this.  /cc @alyssawilk

Reply:
Yes, the plan is to keep the extension point API hidden for the foreseeable future, so we should hopefully not disrupt users if we need to change things when the unified config xDS client in Envoy becomes a reality.

/cc @lizan for another API maintainer on their thoughts.

Reply:
Lizan approved the PR, so will leave it to Adi for final comments

Reply:
@adisuissa @alyssawilk 
I pushed up a commit that adds a `reset()` function to the `XdsResourcesDelegate`, the purpose of which is to clear the existing resources in the xDS store when there is a change in the resource subscription, before adding the updated resources to the store.  This will enable us to handle updates to resources (update, not clear + add), while also clearing out resources that are no longer subscribed to.

We still need the authority id because as the tests in https://github.com/envoyproxy/envoy/pull/22823 show, it's possible to have two resources of the same type and name coming from two separate xDS servers.  We will need to be able to distinguish the resources to be added/updated/deleted in that situation.

Reply:
I changed this again to remove the `reset()` function, based on Adi astutely pointing out that we could have two or more GrpcMuxImpls for the same authority + type (different ApiConfigSources).  Instead, I think it makes sense to just receive the config updates, and allow the TTL'ing mechanism to handle eviction of stale resources (see https://github.com/envoyproxy/envoy/issues/22303 for details).